### DISCUSSION TITLE
Customized EM field input for ImpactX simulation?

### DISCUSSION THREAD

**Discussion by ycaophysics** [BLAST-ImpactX/impactx#687] [Q&A] [answered]
Hi ImpactX creators. 
  I'm wondering if there's any function in ImpactX that could input an OpenPMD file (or a text file) as an external electro-magnetic field? I think this function could be great addition for the impactX code. (Or is it already implemented and I couldn't find it?)

Thanks
Yuxuan Cao

**Comment by ax3l** [comment, marked as answer]
Hi @ycaophysics,

We do not have this feature in ImpactX yet, but WarpX supports exactly this. If you are looking for a complex, time-based integrator, WarpX could be what you could use already:
https://warpx.readthedocs.io/en/latest/usage/parameters.html#external-fields

For ImpactX, we already started to add on-axis and edge fields to various elements, but not yet 3D versions.
(see [Python](https://impactx.readthedocs.io/en/latest/usage/python.html#lattice-elements) elements or [input files](https://impactx.readthedocs.io/en/latest/usage/parameters.html#lattice-elements))

In the past, we had an openPMD reader in IMPACT-Z for ACE3P field data, which we should also implement in ImpactX again. I'll discuss this in our next developer meeting. If you have a concrete science case that could benefit from this, please feel free to share it here so we can see if we already have a funded activity to prioritize it (or we can help if you/someone wants to contribute it :) ).

**Reply by ycaophysics** [reply]
Hi Dr Huebl: 
 What I want to do is to fine tune the electrons for emittance and energy spread with many many trial simulations for a given electric field in LWFA. And I want to use machine learning (inverse problem or backtracking) to find the optimal condition to suppress emittance and energy spread, and increase the brightness for external beam injections. However, with WarpX's PIC simulation it would take way too long to complete a single simulation (3+hours for a quasi-3D boosted-frame), while I think I need around ~10000 of these trials. I don't know how much it would compare to ImpactX's speed if there's no plasma inside the fields. If both are same order of speed, I think that would be okay.    
 
 Thanks
 Yuxuan Cao

**Reply by ax3l** [reply]
Hi @ycaophysics , sorry for the delay in answering.

In the meantime, we improved details on how to add new elements. If you have a good model and know how to implement it, you can already do it using one of these methods: https://impactx.readthedocs.io/en/latest/usage/howto/add_element.html Maybe the Python programmable element is the way to go for now?
